src/libs/type-utils/lodash-fp-helpers.test.ts

```diff
@@ -0,0 +1,451 @@
+      const handlersWatcher = jest.fn()
+
+      const handler1 = <F extends AnyPromiseFn, P>(
+        arg: string
+      ) => {
+        return createHandlerAsync<P, F>(async (executeInner: () => Promise<P>) => {
+          const result = await executeInner()
+          handlersWatcher('handler1', arg)
+          return result
+        })
+      }
+
+      const handler2 = <F extends AnyPromiseFn, P>(
+        arg1: string,
+        arg2: number
+      ) => {
+        return createHandlerAsync<P, F>(async executeInner => {
+          const result: P = await executeInner()
+          handlersWatcher('handler2', arg1, arg2)
+          return result
+        })
+      }
+
+      const watcher = jest.fn()
+
+      // Act
+      // example with inline mainFn arg
+      const myFn = withHandlers([
+        handler1('hi'),
+        handler2('sup', 7)
+      ],
+      async (a: string, b: string): Promise<string> => {
+        watcher(a, b)
+        await delay(100)
+        return `${a} ${b}!`
+      })
+
+      const myResult = await myFn('hello', 'there')
+
+      // Assert
+      expect(myResult).toBe('hello there!')
+      expect(watcher).toBeCalledTimes(1)
+      expect(watcher).toBeCalledWith('hello', 'there')
+
+      expect(handlersWatcher).toBeCalledTimes(2)
+      expect(handlersWatcher).toHaveBeenNthCalledWith(1, 'handler1', 'hi')
+      expect(handlersWatcher).toHaveBeenNthCalledWith(2, 'handler2', 'sup', 7)
+    })
+
+    it('handles curried handler wrapper promise-fns', async () => {
+      // Arrange
+      const handlersWatcher = jest.fn()
+      const handler1Fn = <F extends AnyPromiseFn, P>(
+        arg: string,
+        fn: GenericPromiseFn<F, P>
+      ): GenericPromiseFn<F, P> => {
+        return async (...args: Parameters<F>): Promise<P> => {
+          // simulate typical pattern of calling fn, then do handling
+          const result = await fn(...args)
+          handlersWatcher('handler1', arg)
+          return result
+        }
+      }
+      const handler1 = _.curry(handler1Fn) as SafeCurry2<typeof handler1Fn>
+
+      const handler2Fn = <F extends AnyPromiseFn, P>(
+        arg1: string,
+        arg2: number,
+        fn: GenericPromiseFn<F, P>
+      ): GenericPromiseFn<F, P> => {
+        return async (...args: Parameters<F>): Promise<P> => {
+          // simulate typical pattern of calling fn, then do handling
+          const result = await fn(...args)
+          handlersWatcher('handler2', arg1, arg2)
+          return result
+        }
+      }
+      const handler2 = _.curry(handler2Fn) as SafeCurry3<typeof handler2Fn>
+
+      const watcher = jest.fn()
+
+      const mainFn = async (a: string, b: string): Promise<string> => {
+        watcher(a, b)
+        await delay(100)
+        return `${a} ${b}!`
+      }
+
+      // Act
+      // testing curried functions for completeness, but note that they have
+      // inherently less type safety compared to using curryOnlyLastArg or createHandlerAsync helpers
```

Comment:
> note that they have inherently less type safety compared to using curryOnlyLastArg or createHandlerAsync helpers

Why? Isn't SafeCurry supposed to provide that type safety?

Reply:
SafeCurryX provides a general improvement to type safety of curried functions.  The challenge with these handler functions is that they are essentially unknown on the function signature they are wrapping... combine that with the allowed permutations of a curried function (none, some or all args given produces different results), and you get TS able to protect against some, but not all type violations. 

That's better then nothing in this scenario... and SafeCurryX does better in other scenarios.  But this is why the alternatives provided here are recommended over just SafeCurryX when dealing with unknown fn wrappers/handlers.

I can articulate that in a comment :)

Reply:
when implementing your other suggestion to move away from exported function overloads to declared overload interfaces for a couple of the key exported functions in lodash-fp-helpers.ts the type safety on SafeCurryX + withHandlers improved greatly!  So I was able to replace this comment with a more positive one.